mllib/src/main/scala/org/apache/spark/ml/tree/impl/PartitionInfo.scala

```diff
@@ -0,0 +1,422 @@
+   *       Sort corresponding range of instances based on bit vector.
+   * Update nodeOffsets, activeNodes:
+   *   Split offsets for nodes which split (which can be identified using the bit vector).
+   *
+   * @param instanceBitVector  Bit vector encoding splits for the next level of the tree.
+   *                    These must follow a 2-level ordering, where the first level is by node
+   *                    and the second level is by row index.
+   *                    bitVector(i) = false iff instance i goes to the left child.
+   *                    For instances at inactive (leaf) nodes, the value can be arbitrary.
+   * @return Updated partition info
+   */
+  def update(
+      instanceBitVector: BitSet,
+      newNumNodeOffsets: Int,
+      labels: Array[Byte],
+      metadata: DecisionTreeMetadata): PartitionInfo = {
+    // Create a 2-level representation of the new nodeOffsets (to be flattened).
+    // These 2 levels correspond to original nodes and their children (if split).
+    val newNodeOffsets = nodeOffsets.map(Array(_))
+    val newFullImpurityAggs = fullImpurityAggs.map(Array(_))
+
+    val newColumns = columns.zipWithIndex.map { case (col, index) =>
+      index match {
+        case 0 => first(col, instanceBitVector, metadata,
+          labels, newNodeOffsets, newFullImpurityAggs)
+        case _ => rest(col, instanceBitVector, newNodeOffsets)
+      }
+      col
+    }
+
+    // Identify the new activeNodes based on the 2-level representation of the new nodeOffsets.
+    val newActiveNodes = new BitSet(newNumNodeOffsets - 1)
+    var newNodeOffsetsIdx = 0
+    var i = 0
+    while (i < newNodeOffsets.length) {
+      val offsets = newNodeOffsets(i)
+      if (offsets.length == 2) {
+        newActiveNodes.set(newNodeOffsetsIdx)
+        newActiveNodes.set(newNodeOffsetsIdx + 1)
+        newNodeOffsetsIdx += 2
+      } else {
+        newNodeOffsetsIdx += 1
+      }
+      i += 1
+    }
+    PartitionInfo(newColumns, newNodeOffsets.flatten, newActiveNodes, newFullImpurityAggs.flatten)
+  }
+
+
+  /**
+   * Sort the very first column in the [[PartitionInfo.columns]]. While
+   * we sort the column, we also update [[PartitionInfo.nodeOffsets]]
+   * (by modifying @param newNodeOffsets) and [[PartitionInfo.fullImpurityAggs]]
+   * (by modifying @param newFullImpurityAggs).
+   *
+   * @param col The very first column in [[PartitionInfo.columns]]
+   * @param metadata Used to create new [[ImpurityAggregatorSingle]] for a new child
+   *                 node in the tree
+   * @param labels   Labels are read as we sort column to populate stats for each
+   *                 new ImpurityAggregatorSingle
+   */
+  private def first(
+      col: FeatureVector,
+      instanceBitVector: BitSet,
+      metadata: DecisionTreeMetadata,
+      labels: Array[Byte],
+      newNodeOffsets: Array[Array[Int]],
+      newFullImpurityAggs: Array[Array[ImpurityAggregatorSingle]]) = {
+    activeNodes.iterator.foreach { nodeIdx =>
```

Comment:
activeNodes will be replaced by an Array of nodes.  Some will be leafs (no split), and some will not.  Could still use instanceBitVector to avoid recomputing split decisions for each row.